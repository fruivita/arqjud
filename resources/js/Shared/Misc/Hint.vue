<!--
    Exibe informação sobre determinado elemento no mouseover e mouseout.

    @link https://vuejs.org/guide/introduction.html
    @link https://tailwindcss.com/docs
    @link https://inertiajs.com/
 -->

<script setup>
import { arrow, computePosition, flip, offset, shift } from '@floating-ui/dom';
import { onClickOutside } from '@vueuse/core';
import { isArray } from 'lodash';
import { ref } from 'vue';

const props = defineProps({
    texto: { type: [Array, String], required: true },
    posicao: {
        type: String,
        default: 'bottom',
        validator(value) {
            return ['bottom', 'top', 'left', 'right'].includes(value);
        },
    },
});

const referenceRef = ref(null);
const floatingRef = ref(null);
const arrowRef = ref(null);
const exibirHint = ref(false);

const mouveover = () => {
    if (exibirHint.value !== true) {
        exibirHint.value = true;
        calcularPosicao();
    }
};

const mouseout = () => {
    exibirHint.value = false;
};

const calcularPosicao = async () => {
    const { x, y, middlewareData, placement } = await computePosition(
        referenceRef.value,
        floatingRef.value,
        {
            placement: props.posicao.value,
            middleware: [offset(8), flip(), shift(), arrow({ element: arrowRef.value })],
        }
    );

    Object.assign(floatingRef.value.style, {
        left: `${x}px`,
        top: `${y}px`,
    });

    const { x: arrowX, y: arrowY } = middlewareData.arrow;

    const ladoOposto = {
        left: 'right',
        right: 'left',
        bottom: 'top',
        top: 'bottom',
    }[placement.split('-')[0]];

    Object.assign(arrowRef.value.style, {
        left: arrowX ? `${arrowX}px` : '',
        top: arrowY ? `${arrowY}px` : '',
        bottom: '',
        right: '',
        [ladoOposto]: '-4px',
    });
};

onClickOutside(floatingRef, () => {
    exibirHint.value = false;
});
</script>

<template>
    <div class="inline-block">
        <div
            ref="referenceRef"
            @mouseout="mouseout"
            @mouseover="mouveover"
            class="flex"
            dusk="toggle"
        >
            <slot></slot>
        </div>

        <div
            v-show="exibirHint"
            ref="floatingRef"
            class="absolute top-0 left-0 z-10 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50"
            dusk="hint"
        >
            <template v-if="isArray(texto)">
                <span v-for="(txt, indice) in texto" :key="indice" class="block">{{ txt }}</span>
            </template>

            <template v-else>
                <span>{{ texto }}</span>
            </template>

            <div
                ref="arrowRef"
                class="absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600"
            ></div>
        </div>
    </div>
</template>
