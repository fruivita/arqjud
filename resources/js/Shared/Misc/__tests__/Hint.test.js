/**
 * Testes para o componente Hint.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import Hint from '@/Shared/Misc/Hint.vue';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';

let mountFunction;

beforeEach(() => {
    mountFunction = (options = {}) => {
        return mount(Hint, { ...options });
    };
});

describe('Hint', () => {
    // Caminho feliz
    const posicoes = ['bottom', 'top', 'left', 'right'];

    test('aceita apenas as posições definidas', () => {
        const validator = Hint.props.posicao.validator;

        posicoes.forEach((type) => expect(validator(type)).toBe(true));
        expect(validator('foo')).toBe(false);
    });

    test('propriedades do componente estão definidas', () => {
        expect(Hint.props).toMatchObject({
            texto: { type: [Array, String], required: true },
            posicao: { type: String, default: 'bottom' },
        });
    });

    test('renderiza o componente default respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { texto: 'foo' },
                slots: { default: 'loren' },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza o componente informando um array de textos respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: { texto: ['foo', 'bar', 'baz'] },
                slots: { default: 'loren' },
            }).html()
        ).toMatchSnapshot();
    });

    test.each(posicoes)(
        'renderiza o componente nas posições específicas respeitando o snapshot',
        (posicao) => {
            expect(
                mountFunction({
                    props: { texto: 'foo', posicao: posicao },
                    slots: { default: 'loren' },
                }).html()
            ).toMatchSnapshot();
        }
    );

    test('alterna a visibilidade do hint nos eventos mouseover e mouseout no botão', async () => {
        const wrapper = mountFunction({
            props: { texto: 'foo' },
        });

        const hint = wrapper.get('[dusk="hint"]');

        expect(hint.isVisible()).toBeFalsy();

        await wrapper.get('[dusk="toggle"]').trigger('mouseover');

        expect(hint.isVisible()).toBeTruthy();

        await wrapper.get('[dusk="toggle"]').trigger('mouseout');

        expect(hint.isVisible()).toBeFalsy();
    });
});
