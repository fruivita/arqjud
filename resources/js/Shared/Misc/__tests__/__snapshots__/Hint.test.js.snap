// Vitest Snapshot v1

exports[`Hint > renderiza o componente default respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\">
  <div class=\\"flex\\" dusk=\\"toggle\\">loren</div>
  <div class=\\"absolute top-0 left-0 z-10 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"hint\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Hint > renderiza o componente informando um array de textos respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\">
  <div class=\\"flex\\" dusk=\\"toggle\\">loren</div>
  <div class=\\"absolute top-0 left-0 z-10 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"hint\\" style=\\"display: none;\\"><span class=\\"block\\">foo</span><span class=\\"block\\">bar</span><span class=\\"block\\">baz</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Hint > renderiza o componente nas posições específicas respeitando o snapshot 1`] = `
"<div class=\\"inline-block\\">
  <div class=\\"flex\\" dusk=\\"toggle\\">loren</div>
  <div class=\\"absolute top-0 left-0 z-10 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"hint\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Hint > renderiza o componente nas posições específicas respeitando o snapshot 2`] = `
"<div class=\\"inline-block\\">
  <div class=\\"flex\\" dusk=\\"toggle\\">loren</div>
  <div class=\\"absolute top-0 left-0 z-10 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"hint\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Hint > renderiza o componente nas posições específicas respeitando o snapshot 3`] = `
"<div class=\\"inline-block\\">
  <div class=\\"flex\\" dusk=\\"toggle\\">loren</div>
  <div class=\\"absolute top-0 left-0 z-10 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"hint\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;

exports[`Hint > renderiza o componente nas posições específicas respeitando o snapshot 4`] = `
"<div class=\\"inline-block\\">
  <div class=\\"flex\\" dusk=\\"toggle\\">loren</div>
  <div class=\\"absolute top-0 left-0 z-10 cursor-default rounded-md bg-primaria-600 px-3 py-1.5 text-sm font-bold text-primaria-50 dark:bg-secundaria-600 dark:text-secundaria-50\\" dusk=\\"hint\\" style=\\"display: none;\\"><span>foo</span>
    <div class=\\"absolute h-[8px] w-[8px] rotate-45 bg-primaria-600 dark:bg-secundaria-600\\"></div>
  </div>
</div>"
`;
